doc/continuous_integration/dockerfile.md

```diff
@@ -0,0 +1,93 @@
+
+## Cacher Stage
+
+A `cacher` stage is then started to gather together the necessary source files to eventually build. A directory for the underlay workspace is then created and populated using `vcs` with the respective `.repos` file that defines the relevant repositories to pull and particular versions to checkout into the source directory. More info on vcstool can be found here: 
+
+* [vcstool](https://github.com/dirk-thomas/vcstool)
+  * CLI for working with multiple repositories easier
+
+The `.repos` file is not copied directly into the `src` folder to avoid any restructuring of the yaml data from unintentionally busting the docker build cache in later stages. The ephemeral files within `.git` repo folders are similarly removed and help bolster deterministic builds. 
+
+The overlay workspace is then also created and populated using all the files in the docker build context, i.e. the [root](/) directory of the repo. This is done after the underlay is cloned to avoid having to re-download underlay source files if the `.repos` files are unchanged. However, if the `.repos` file is changed, and different source files are cloned, this can then bust the docker build cache. Other ephemeral or unessential project files are safely ignored using the [`.dockerignore`](/.dockerignore) config. If ever the docker build cache is cache is somehow stale, using the docker build flag `--no-cache` may be used to freshly build anew.
+
+Finally the `cacher` stage copies all manifest related files in place within the `/opt` directory into a temporary mirrored directory that later stages can copy from without unnecessarily busting it's docker build cache. The [`source.Dockerfile`](/.dockerhub/source.Dockerfile) provides an advance example of avoiding ignored packages, or packages that are unnecessary as overlay dependencies.
+
+## Builder Stage
+
+A `builder` stage is then started to install external dependencies and compile the respective workspaces. Static CI dependencies are first installed before any later potential cache busting directives. These include:
+
+* [ccache](https://ccache.dev)
+  * Compiler cache for speeding up recompilation
+* [lcov](http://ltp.sourceforge.net/coverage/lcov.php)
+  * Front-end for GCC's coverage testing tool gcov
+
+### Install Dependencies
+
+Dependencies for the underlay workspace are then installed using `rosdep` by pointing to the manifest files within the mirrored source directory copied from the `cacher` stage. This ensures the lengthy process of downloading, unpacking, and installing any external dependencies can be skipped using the docker build cache as long as the manifest files within the underlay remain unchanged. More info on rosdep can be found here:
```

Comment:
We want to avoid breaking the docker build cache that includes the image layer where rosdep installed all the build/run/test dependencies; even if the source files change but the package manifest don't. This is done by copying the manifest files that the cacher stage duplicated in a temp directory, rather than the entire src directory directly from the docker build context that could break the docker build cache.